### ISSUE TITLE
TSX type inference is broken when using intersection types

### ISSUE THREAD

**Issue by iyegoroff** [Microsoft/TypeScript#21632]
**TypeScript Version:**  2.7.1, 2.8.0-dev.20180204

**Search Terms:** jsx tsx intersection

**Code**

```ts
import * as React from 'react';

type Props<T> = T & {prop: string};

class Component<T = {valid: string}> extends React.Component<Props<T>> {}

<>
  <Component prop={''} invalid={''} />
  <Component prop={''} valid={''} invalid={''}/>
</>
```

**Expected behavior:**
Should generate 2 errors as in 2.6.2:
```
main.tsx(8,14): error TS2322: Type '{ prop: ""; invalid: ""; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<{ valid: string; }>> & Readonly<{ childr...'.
  Type '{ prop: ""; invalid: ""; }' is not assignable to type 'Readonly<Props<{ valid: string; }>>'.
    Property 'valid' is missing in type '{ prop: ""; invalid: ""; }'.
main.tsx(9,35): error TS2339: Property 'invalid' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<{ valid: string; }>> & Readonly<{ childr...'.
```
**Actual behavior:**
No errors are generated

**Playground Link:** -

**Related Issues:** maybe related #21427

**Comment by RyanCavanaugh** [comment]
@weswigham thoughts on this?

**Comment by weswigham** [comment]
I think the jsx inference/contextual typing PR I currently have open is mostly to fix exactly this, since I discovered the same issue when looking at changing formik.

**Comment by weswigham** [comment]
Also, hold on: those 2.6.2 errors are themselves erroneous. Setting a default doesn't constraint the type variable - literally any props _should be_ be OK to pass for that component - we just messed up before and instantiated with the default like it was the constraint before.

**Comment by RyanCavanaugh** [comment]
>  Setting a default doesn't constraint the type variable - literally any props should be be OK to pass for that component

This was my interpretation as well. Seems useless to have a generic component if you're not allowed to infer a type from its props.

**Comment by iyegoroff** [comment]
@weswigham, so the actual behavior in 2.7.1 is correct and it will stay like this after your PR will be merged?

**Comment by weswigham** [comment]
Yeah. You should be using `extends` in that type definition if you actually wanted that field required.

**Comment by iyegoroff** [comment]
OK, thanks for clarification.
